### PR TITLE
FIX: Refactor GLM results for memory efficiency and core API stability (#2381)

### LINKED ISSUE #2381: Simplify the LikelihoodModelResults.
Stores un-necessary stuff (input data, residuals), hence memory-consuming.
Partially handled in nilearn/nistats#72, but full removal should be considered.

### PR DESCRIPTION
**Pull request by BabaSanfour**
Refactors GLM results to improve memory efficiency by eliminating redundant storage of large data arrays (input images, whitened data, and residuals) within the `RegressionResults` objects.

### Key Changes
- **Memory Optimization**: `RegressionResults` no longer stores `Y`, `whitened_Y`, or `residuals` by default. These are now computed or retrieved via methods only when explicitly requested.
- **Summary Statistics**: Common statistics like `SSE`, `r_square` and `MSE` are now computed once during `fit()` and stored as pre-calculated static attributes, maintaining fast performance for most use cases (like maps and summaries).

### API Evolution
Most voxel-wise attributes have transitioned from attributes/properties to methods.

#### Residuals and Predictions
- **Before**: 
  ```python
  res = model.residuals    # Attribute (High memory)
  pred = model.predicted   # Attribute
  ```
- **After**: 
  ```python
  res = model.residuals(Y)  # Method requiring input data 'Y'
  pred = model.predicted()  # Method call
  ```

#### Summary Statistics (R-squared)
- **Before**: 
  ```python
  r2 = model.r_square      # Property
  ```
- **After**: 
  ```python
  r2 = model.r_square()    # Method call (returns pre-calculated map)
  ```

#### Normalized Residuals
- **Before**: 
  ```python
  norm_res = model.normalized_residuals  # Attribute
  ```
- **After**: 
  ```python
  norm_res = model.normalized_residuals(Y)  # Method requiring input data 'Y'
  ```

### Verification Results
- **Code Quality**: Full cleanup with `ruff` and `pre-commit` validation across all modified files.
- **Test Suite**: 304 non-BIDS GLM tests passing (Regression, Model, First/Second Level, Reporter).

### Next Step Before Merging
- **Phase 2**: Merge and eventually deprecate `SimpleRegressionResults`. The new `RegressionResults` is now efficient enough to serve both standard and memory-minimized use cases.
- **Phase 3**: Audit and update `io.py` and BIDS saving functions. 

> [!NOTE]
> Currently, tests for `io.py` and some edge cases with `minimize_memory=True` (specifically for late-stage statistical map saving) are expected to fail. I will work on resolving these while addressing Phases 2 and 3.

### Linked Issues
- Fixes #2381 / Potentially also #5588
- Might help with #2421

### CODE REVIEW

**Comment by BabaSanfour** [comment]
- I was checking the rest of the open issues in the Kanban with no status ( #5588 ) and I found this other issue that I think is directly related to the proposed fixes above. I reanalyzed the code and thought of 3 potential suggestions. I will better write these tomorrow and then maybe we can have another discussion @emdupre @bthirion , maybe @Remi-Gau also as they opened the last issue.

**Comment by bthirion** [comment]
Note that this is an API change, so we'll need to go through deprecation cycles etc.

**Comment by BabaSanfour** [comment]
#### Why the memory explodes? 

@emdupre, @bthirion, as mentioned [here](https://neurostars.org/t/nilearn-firstlevelmodel-fit-ram-memory-explosion/34009/11) (#5588), processing large datasets (e.g., shape `(64, 76, 64, 6804)`) causes Nilearn to "eat through 32GB of RAM". I think this is primarily due to a "multi-copy" regression result structure. 

For an array of that shape at `float64`, a single copy is **17 GB**. The current implementation creates multiple redundant copies:

1.  **Duplicate Input**: `LikelihoodModelResults` saves the data (`Y`) in the object, besides the copy already in the workspace or on disk.
2.  **Redundant Result Copies** in `RegressionResults`:
    *   **Initialization**: Both `self.whitened_Y` and `self.whitened_residuals` are stored immediately.
    *   **Caching**: Usage of `@auto_attr` permanently caches `self.residuals` and `self.normalized_residuals` in RAM.

If the initial input $Y$ is **1 GB**, the current implementation ballooned to **5 GB** of persistent RAM usage. For the 17 GB dataset cited above, the overhead reaches **85 GB**.

---

### What could work?

1.  **Option 1: The Clean & Safe Refactor (This PR)**: 
    *   Move from attributes to methods for all data-dependent results.
    *   100% memory efficiency; aligns with `scikit-learn` API (estimators don't store $Y$ nor transformation of $Y$).
2.  **Option 2: The Minimum Change**: 
    *   Only drop `Y` from being saved in `LikelihoodModelResults`.
    *   Only addresses 20% of the total bloat. Maybe more if we find options to remove `self.whitened_Y` and `self.whitened_residuals`. But we are _always_ still left with `Y`, `self. residuals` and `self. normalized_residuals`
3.  **Option 3: A Hybrid Approach**: 
    *   Add a `return_residuals` flag to `fit()`. If `True`, it computes and stores residuals as attributes (`self. residuals_` and `self.normalized_residuals_` using `self. residuals(Y)` and `self.normalized_residuals(Y)` ; otherwise, they stay as methods and user could compute them using `model.residuals(Y)`. Eventually we can have a `return_normalized_residuals` flag too.   
    *   This keeps the option for users who want "everything in a box".

Although I more fan of **Option 1**, I think **Option 3** could eventually be the way to go. It also allows us to eventually **merge `RegressionResults` and `SimpleRegressionResults`** into a single, unified logic for handling GLM results and removes the need for manual memory workarounds.

For the API change, I can add deprecations into account and take care of that for the next 2/3 releases.

File: nilearn/glm/second_level/second_level.py [hunk 1]
```diff
@@ -854,10 +858,10 @@ def _get_element_wise_model_attribute(
         """
         check_is_fitted(self)
         # check if valid attribute is being accessed.
-        all_attributes = dict(vars(RegressionResults)).keys()
         possible_attributes = [
-            prop for prop in all_attributes if "__" not in prop
+            prop for prop in dict(vars(RegressionResults)) if "__" not in prop
         ]
+        possible_attributes += RegressionResults.__static_attributes__
```
Comment (emdupre) [inline nilearn/glm/second_level/second_level.py:934]: I believe `__static_attributes__` was only added recently (maybe [as recently as 3.13](https://docs.python.org/fr/3/reference/datamodel.html#type.__static_attributes__)) ; we support back to 3.10, though, so we would need to use a more general method

Comment (emdupre) [reply nilearn/glm/second_level/second_level.py:934]: Could we revert to the previous formulation here ?

Comment (BabaSanfour) [reply nilearn/glm/second_level/second_level.py:934]: I changed this to avoid some tests that were crushing as we changed from attributes to methods. I could find a way around to go back to the prev formulation. This will be changed eventually depending on the Option we decide to go with.

**Comment by bthirion** [comment]
> #### Why the memory explodes?
> 
> @emdupre, @bthirion, as mentioned [here](https://neurostars.org/t/nilearn-firstlevelmodel-fit-ram-memory-explosion/34009/11) (#5588), processing large datasets (e.g., shape `(64, 76, 64, 6804)`) causes Nilearn to "eat through 32GB of RAM". I think this is primarily due to a "multi-copy" regression result structure.
> 
> For an array of that shape at `float64`, a single copy is **17 GB**. The current implementation creates multiple redundant copies:
> 
>     1. **Duplicate Input**: `LikelihoodModelResults` saves the data (`Y`) in the object, besides the copy already in the workspace or on disk.
> 
>     2. **Redundant Result Copies** in `RegressionResults`:
>        
>        * **Initialization**: Both `self.whitened_Y` and `self.whitened_residuals` are stored immediately.
>        * **Caching**: Usage of `@auto_attr` permanently caches `self.residuals` and `self.normalized_residuals` in RAM.
> 
> 
> If the initial input Y is **1 GB**, the current implementation ballooned to **5 GB** of persistent RAM usage. For the 17 GB dataset cited above, the overhead reaches **85 GB**.
> ### What could work?
> 
>     1. **Option 1: The Clean & Safe Refactor (This PR)**:
>        
>        * Move from attributes to methods for all data-dependent results.
>        * 100% memory efficiency; aligns with `scikit-learn` API (estimators don't store 
>            Y
>           nor transformation of 
>            Y
>          ).
> 
>     2. **Option 2: The Minimum Change**:
>        
>        * Only drop `Y` from being saved in `LikelihoodModelResults`.
>        * Only addresses 20% of the total bloat. Maybe more if we find options to remove `self.whitened_Y` and `self.whitened_residuals`. But we are _always_ still left with `Y`, `self. residuals` and `self. normalized_residuals`
> 
>     3. **Option 3: A Hybrid Approach**:
>        
>        * Add a `return_residuals` flag to `fit()`. If `True`, it computes and stores residuals as attributes (`self. residuals_` and `self.normalized_residuals_` using `self. residuals(Y)` and `self.normalized_residuals(Y)` ; otherwise, they stay as methods and user could compute them using `model.residuals(Y)`. Eventually we can have a `return_normalized_residuals` flag too.
>        * This keeps the option for users who want "everything in a box".
> 
> 
> Although I more fan of **Option 1**, I think **Option 3** could eventually be the way to go. It also allows us to eventually **merge `RegressionResults` and `SimpleRegressionResults`** into a single, unified logic for handling GLM results and removes the need for manual memory workarounds.
> 
> For the API change, I can add deprecations into account and take care of that for the next 2/3 releases.

Thx for the detailed analysis !
* Storing `whitened residuals`, `residuals` and `normalized residuals` seems overly redundant. If we could keep only one among 3, this would be beneficial.
* option 3 sounds the right approach. Just, the sklearn API requires to provide these flags at instantiation time, not fit time, but this is a detail.
This being said, I'll be super happy to see this happen !
